Scenario: 'Visitor can navigate the site'

raptorf1 opened the pull request:
[PT Story](https://www.pivotaltracker.com/story/show/165594709)

# User Story
```
As a visitor,
In order to navigate the site
I would like to use a navbar
```

### What we worked on in this feature
* Adds feature test with 10 steps
* Creates 2 controllers :products & :about_us
* Creates a separate HAML file for the NavBar and renders it to the application view
* Adds the Tailwind CSS framework for the whole application

### What we learned
* Git-ponging between 3 people
* Connection between views (rendering the _navigation partial file)
* HAML syntax

### Screenshots
![Screenshot from 2019-05-02 18-47-21](https://user-images.githubusercontent.com/48210172/57092066-dad4f200-6d0a-11e9-9a43-fd2f3a72b663.png)


![Screenshot from 2019-05-02 18-47-24](https://user-images.githubusercontent.com/48210172/57092123-ee805880-6d0a-11e9-8120-881bbe037775.png)

Carrosen committed:
creates feature test for 'visitor can navigate the site' scenario
creates products controller
sets product controller as root_path
first step of feature, going green

boamatule committed:
added text in the layout application view,
added basic_step.rb code.
second step of scenario is going green.

raptorf1 committed:
Adds 'MENU' link to the application view. Link directing to root_path. 3rd step of scenario green.

Carrosen committed:
adds opening hours to products view, 4th step of scenario going green

boamatule committed:
added 'park location' text on Products view, 5th test goes green.

raptorf1 committed:
Creates controller 'About_us', makes the 'ABOUT US' text on the application view a link to redirect to the new controller, 6th step of scenario green.

Carrosen committed:
adds text 'THE KIMCHEN FOODTRUCK' to about us view, added code to assertion_steps.rb, all tests going green

raptorf1 committed:
Creates NavBar haml file, converts NavBar to list, renders NavBar into the application view, adds Tailwind CSS framework.

faraznaeem commented on app/controllers/about_us_controller.rb:
```diff
@@ -0,0 +1,4 @@
+class AboutUsController < ApplicationController
+  def index
+  end
+end
```
Why is this controller needed?

raptorf1 replied:
The way we designed the LO-FI's we have an 'About Us' section we would like to use for the history let's say of the restaurant.
Since this feature is about navigating the site we wrote steps that require the user to click on the 'About Us' link and back.
To have an 'About Us' view then we must have an appropriate controller. Simple one with index only. So as to navigate to the page.

faraznaeem commented on config/routes.rb:
```diff
@@ -1,3 +1,6 @@
+  get 'about_us/index'
+  get 'products/index'
+
+  root "products#index"
```
This is not how we write routes

Carrosen replied:
true

raptorf1 replied:
OK. Rewrote them

faraznaeem commented on features/visitor_can_navigate_the_site.feature:
```diff
@@ -0,0 +1,23 @@
+
+
+    
+
```
Delete empty lines

Carrosen replied:
on it

raptorf1 replied:
Removed

faraznaeem commented:
Please add a screenshot as well

oliverochman commented on features/visitor_can_navigate_the_site.feature:
```diff
@@ -0,0 +1,23 @@
+
+
+    
+
```
remove empty lines

Carrosen replied:
on it

raptorf1 replied:
removed

oliverochman commented on features/step_definitions/basic_steps.rb:
```diff
@@ -0,0 +1,11 @@
+  visit root_path
+end
+  
+Then("I should see {string}") do |content|
```
add this step to the assertions steps instead

Carrosen replied:
on it

raptorf1 replied:
Also added.

oliverochman commented on features/step_definitions/basic_steps.rb:
```diff
@@ -0,0 +1,11 @@
+  expect(page).to have_content content 
+end
+  
+Then("I click {string} link") do |link|
```
add this step to the assertions steps instead

Carrosen replied:
on it

raptorf1 replied:
OK. Added ;)

oliverochman commented on features/visitor_can_navigate_the_site.feature:
```diff
@@ -0,0 +1,23 @@
+        And I should see "MENU"
+        And I should see "Opening hours"
+        And I should see "Park location"
+        Then I click "ABOUT US" link
```
This should be "When"

raptorf1 replied:
OK. Changed to 'When'. Edited the cucumber steps also.

oliverochman commented on config/routes.rb:
```diff
@@ -1,3 +1,6 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  get 'about_us/index'
```
update the way you define these routes

e..g.
`root controller: :products, action: :index`
`resources :about_us, only: [:index]`

raptorf1 replied:
Thank you.
Updated ;)

oliverochman commented on app/controllers/about_us_controller.rb:
```diff
@@ -0,0 +1,4 @@
+class AboutUsController < ApplicationController
```
I would say that adding an about page in this feature is out of scope

Carrosen replied:
Okay, but without navigation, why is there a feature about this? Since it's about navigation. I don't really get what this feature is without a second page?

faraznaeem replied:
You don't need a controller to navigate to different pages, you can reuse the products controller. 
https://stackoverflow.com/questions/18703349/how-to-add-a-new-action-to-the-existing-controller

That being said why do you start a story about navigating if there is nothing to navigate to?

Carrosen replied:
ahhh nice. I thought you did need one. That clears things up.

There is something to navigate to, to about us.

oliverochman commented on app/views/layouts/application.html.haml:
```diff
@@ -2,9 +2,11 @@
     = csrf_meta_tags
     = stylesheet_link_tag    'application', media: 'all', data: {'turbolinks-track' => 'reload'}
     = javascript_include_tag 'application', data: {'turbolinks-track' => 'reload'}
+    %link(href="https://cdn.jsdelivr.net/npm/tailwindcss/dist/tailwind.min.css" rel="stylesheet")
```
Adding tailwind in this story is out of scope

Carrosen replied:
true

Carrosen replied:
done

raptorf1 committed:
Addresses GitHub PR comments. Redefines routes, removes Tailwind, removes empty lines from feature file, added assertion steps to the appropriate file.

raptorf1 commented:
> Please add a screenshot as well

Screenshots added.

boamatule committed:
deleted about_us controller
created route for the about_us page through the products controller
moved navigation to the newly created folder named Partials.

tochman commented on app/controllers/products_controller.rb:
```diff
@@ -0,0 +1,8 @@
+  end
+
+  def about_us
+  end
```
😳Why is this method defined?

raptorf1 replied:
To navigate to another page based on our design sprint. After Oliver's recommendation the action is removed.

tochman commented on config/routes.rb:
```diff
@@ -1,3 +1,6 @@
+
+  root controller: :products, action: :index
+  resources :products, only: [:index]
+  get "products/about_us"
```
Why is this route defined?

raptorf1 replied:
Look my reply above.

tochman commented on features/step_definitions/assertion_steps.rb:
```diff
@@ -0,0 +1,11 @@
+
+Then("I should see {string}") do |content|
+  expect(page).to have_content content 
+end
```
Assertions should be placed in a separate file

raptorf1 replied:
You are in the assertion_steps.rb
